services/lambda-label-successful-prs/CIBot.py

```diff
@@ -0,0 +1,199 @@
+        return all_checks
+
+    def _is_successful(self, pr):
+        """
+        This method checks if all the pipelines corresponding to a latest build of the PR are successful
+        :param pr PullRequest object
+        """
+        # get jenkins object
+        jenkins_obj = self._get_jenkins_obj()
+
+        # get PR number
+        pr_number = pr.number
+
+        # get a list of checks
+        # Github says "status checks"
+        # Jenkins refers to it as a job name
+        checks = self._get_checks()
+
+        # iterate over the checks
+        for check in checks:
+            # Jenkins Job name
+            name = "mxnet-validation/" + check + "/PR-" + str(pr_number)
+
+            try:
+                # get Jenkins job specific object
+                job = jenkins_obj[name]
+            except Exception:
+                logging.error(f'Unknown Job Exception')
+                self.prs_unknown_job += 1
+                return False
+            # get status of the latest build of the job
+            try:
+                latest_build = job.get_last_build()
+                status = latest_build.get_status()
+                logging.info(f'Latest build status for PR {pr_number} is {status}')
+                if(status.upper() == 'FAILURE'):
+                    self.prs_status_fail += 1
+                    return False
+            except Exception:
+                logging.error(f'No Build Exception')
+                self.prs_no_build += 1
+                return False
+        logging.info(f'All 11 builds succeeded! Happy case!')
+        return True
+
+    def _add_label(self, pr):
+        try:
+            pr.add_to_labels('pr-awaiting-review')
+        except Exception:
+            logging.error('Unable to add label')
+
+        # verify that label has been correctly added
+        if(self._check_awaiting_review_label(pr)):
+            logging.info(f'Successfully labeled {pr.number}')
+            self.prs_labeled += 1
+        return
+
+    def _check_awaiting_review_label(self, pr):
+        """
+        This method returns True if pr-awaiting-review label found in PR labels
+        """
+        labels = pr.get_labels()
+        for label in labels:
+            if 'pr-awaiting-review' == label.name:
+                return True
+        return False
+
+    def _check_status_and_label_if_successful(self, pr):
+        """
+        This method checks the CI status of each PR
+        and it labels if the status is successful
+        """
+        # check if the open PR already has a pr-awaiting-review label
+        if(self._check_awaiting_review_label(pr)):
+            logging.info(f'PR {pr.number} already contains the label pr-awaiting-review')
+            self.prs_already_labeled += 1
+            return
+
+        # check the status of the PR and add label if successful
+        if(self._is_successful(pr)):
+            self._add_label(pr)
+        return
```

Comment:
I think this will need to also take into account whether a PR is reviewed or not. For example, when I'm done reviewing a PR with "pr-awaiting-review", I may request change and remove that label, or change it to "pr-awaiting-response".

Reply:
Yeah I'd certainly refrain from using the "pr-" labels but introducing new "ci-" labels. The former are manual labels while the latter are automated ones - don't mix them.

Reply:
as long as the state machine is properly set up, it shouldn't matter whether a label was manually changed or not. Overall, I think labeling PRs with a status should be a fully automated